### PR TITLE
Port pipeline to DSL2

### PR DESCRIPTION
**Pull request by ggabernet**
This PR contains quite some updates, the most important are:

- porting pipeline to DSL2
- Adding further UMI barcode options
- Adding Rmarkdown results report
- Adding full size tests

The rest are listed in the Changelog.

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Make sure your code lints (`nf-core lint .`).
- [x] Documentation in `docs` is updated
- [x] `CHANGELOG.md` is updated
- [x] `README.md` is updated

**Learn more about contributing:** https://github.com/nf-core/bcellmagic/tree/master/.github/CONTRIBUTING.md

### CODE REVIEW

**Comment by apeltzer** [comment]
This is massive 👀

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -20,28 +23,11 @@ jobs:
     strategy:
       matrix:
         # Nextflow versions: check pipeline minimum and current latest
-        nxf_ver: ['20.04.0', '']
+        nxf_ver: ['21.04.0', '21.03.0-edge']
```
Comment (apeltzer) [inline .github/workflows/ci.yml:26]:
```suggestion
        nxf_ver: ['21.04.1', '21.03.0-edge']
```

Could all go to latest stable 21.04.1 with bugfixes

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -66,28 +49,11 @@ jobs:
     strategy:
       matrix:
         # Nextflow versions: check pipeline minimum and current latest
-        nxf_ver: ['20.04.0', '']
+        nxf_ver: ['20.10.0', '']
```
Comment (apeltzer) [inline .github/workflows/ci.yml:52]:
```suggestion
        nxf_ver: ['21.04.1', '']
```

to make it match the latest version as defined above maybe?

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,35 +3,92 @@
...
+
+### :warning: Major enhancements
+
+* The pipeline has been ported to Nextflow DSL2
+* Analysis of TCR repertoires is now also supported
+* Added Rmarkdown report with a summary of the repertoire analysis
+* Analysis of data with RACE 5' protocol is now supported
+* Scripts to download and build new versions of the IMGT database have been added
+* Improvement of UMI handling and possibility to exchange position of C and V-primers
+* Updated to new versions of Immcantation Framework, giving support to AIRR format
```
Comment (apeltzer) [inline CHANGELOG.md:16]:
```suggestion
* Updated to new versions of Immcantation Framework, providing support for the AIRR format
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -3,35 +3,92 @@
...
 * [#85](https://github.com/nf-core/bcellmagic/pull/85): Added primer handling params: `--vprimer_start`, `--cprimer_start`, `--primer_mask_mode`, `--primer_maxeror`, `--primer_consensus`
-* Template update to nf-core tools v1.12.1
 * [#85](https://github.com/nf-core/bcellmagic/pull/85): Added support for TCR data with params: `--loci`
 * [#85](https://github.com/nf-core/bcellmagic/pull/85): Added support for mice data and possibility for other species with params: `--species`
-* [#85](https://github.com/nf-core/bcellmagic/pull/85): Added support for 5' RACE technology with params: `--race_5prime`, `--race_linker`
+* [#85](https://github.com/nf-core/bcellmagic/pull/85): Added support for 5' RACE technology with params: `--protocol`, `--race_linker`
 * [#87](https://github.com/nf-core/bcellmagic/pull/87): Added tests for TCR data
+* [#102](https://github.com/nf-core/bcellmagic/pull/102): Added param `--protocol`.
+* [#102](https://github.com/nf-core/bcellmagic/pull/102): Added support for C-primer in any R1 or R2 with param `--cprimer_position`.
+* [#102](https://github.com/nf-core/bcellmagic/pull/102): Bump versions to 2.0 and NXF 21.04.0.
```
Comment (apeltzer) [inline CHANGELOG.md:30]:
```suggestion
* [#102](https://github.com/nf-core/bcellmagic/pull/102): Bump versions to 2.0 and NXF 21.04.1.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -3,35 +3,92 @@
...
+Note, since the pipeline is now using Nextflow DSL2, each process will be run with its own [Biocontainer](https://biocontainers.pro/#/registry). This means that on occasion it is entirely possible for the pipeline to be using different versions of the same tool. However, the overall software dependency changes compared to the last release have been listed below for reference.
+
+| Dependency   | Old version | New version |
+|--------------|-------------|-------------|
+| Python       | v3.6.11     | 3.8.0,3.0.3 |
+| markdown     | v3.1.1      |             |
+| muscle       | v3.8.1551   |             |
+| vsearch      | v2.11.1     |             |
+| fastqc       | 0.11.8      | 0.11.9      |
+| multiqc      | 1.9         | 1.9         |
```
Comment (apeltzer) [inline CHANGELOG.md:58]: Could update to latest release here without breaking anything, e.g. multiqc == 1.11

File: README.md [hunk 6]
```diff
@@ -4,36 +4,36 @@
 
 [![GitHub Actions CI Status](https://github.com/nf-core/bcellmagic/workflows/nf-core%20CI/badge.svg)](https://github.com/nf-core/bcellmagic/actions)
 [![GitHub Actions Linting Status](https://github.com/nf-core/bcellmagic/workflows/nf-core%20linting/badge.svg)](https://github.com/nf-core/bcellmagic/actions)
-[![Nextflow](https://img.shields.io/badge/nextflow-%E2%89%A520.04.0-brightgreen.svg)](https://www.nextflow.io/)
+[![Nextflow](https://img.shields.io/badge/nextflow-%E2%89%A521.04.0-brightgreen.svg)](https://www.nextflow.io/)
```
Comment (apeltzer) [inline README.md:7]:
```suggestion
[![Nextflow](https://img.shields.io/badge/nextflow-%E2%89%A521.04.1-brightgreen.svg)](https://www.nextflow.io/)
```

File: assets/nf-core_style.css [hunk 7]
```diff
@@ -0,0 +1,70 @@
+body {
```
Comment (apeltzer) [inline assets/nf-core_style.css:1]: Not sure this belongs in here - looks like this is a remnant from an nf-core/rnaseq push?

File: assets/references.bibtex [hunk 8]
```diff
@@ -0,0 +1,189 @@
+@article{ewels2016multiqc,
```
Comment (apeltzer) [inline assets/references.bibtex:1]: Looks like this is a bibtex file for rnaseq pipeline? Dupradar, Hisat2, STAR aren't used in the workflow at all :-)

File: modules/local/presto/presto_splitseq.nf [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+    output:
+    tuple val(meta), path("*_atleast-2.fasta"), emit: fasta
+    path("*_command_log.txt"), emit: logs
+
+    script:
+    """
+    SplitSeq.py group -s $reads --outname ${meta.id} -f CONSCOUNT --num 2 --fasta > "${meta.id}_command_log.txt"
+    """
+    // TODO: sed -n '1~4s/^@/>/p;2~4p' ${dedupped.baseName}_UMI_R1_R2_atleast-2.fastq > ${dedupped.baseName}_UMI_R1_R2_atleast-2.fasta
```
Comment (FriederikeHanssen) [inline modules/local/presto/presto_splitseq.nf:31]: Is this still needed?

File: modules/local/samplesheet_check.nf [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+    input:
+    path samplesheet
+    
+    output:
+    path '*.tsv'
+
+
+    script:  // This script is bundled with the pipeline, in nf-core/dsltwotest/bin/
+    // TODO: update to check samplesheet
```
Comment (FriederikeHanssen) [inline modules/local/samplesheet_check.nf:26]: is this still needed?

File: modules/local/shazam/shazam_tigger_threshold.nf [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    path("Hamming_distance_threshold.pdf")
+
+    script:
+    def software = getSoftwareName(task.process)
+    if (params.loci == 'ig'){
+        """
+        TIgGER-shazam.R $tab $params.loci $params.threshold_method ${imgt_base}/${params.species}/vdj/imgt_human_IGHV.fasta 
+        Rscript -e "library(shazam); write(x=as.character(packageVersion('shazam')), file='${software}.version.txt')"
+        Rscript -e "library(tigger); write(x=as.character(packageVersion('tigger')), file='tigger.version.txt')"
+
```
Comment (FriederikeHanssen) [inline modules/local/shazam/shazam_tigger_threshold.nf:39]:
```suggestion
```

File: nextflow.config [hunk 12]
```diff
@@ -101,27 +108,59 @@ try {
...
     docker.enabled = true
+    singularity.enabled = false
+    podman.enabled = false
+    shifter.enabled = false
+    charliecloud.enabled = false
     // Avoid this error:
     //   WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
     // Testing this in nf-core after discussion here https://github.com/nf-core/tools/pull/351
     // once this is established and works well, nextflow might implement this behavior as new default.
     docker.runOptions = '-u \$(id -u):\$(id -g)'
```
Comment (FriederikeHanssen) [inline nextflow.config:130]: There was this discussion here https://github.com/nf-core/modules/pull/525 on using docker.userEmulation instead. I don't know if this is common now or not, but just wanted to leave it here for discussion.

File: subworkflows/input_check.nf [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+params.options = [:]
+
+include { SAMPLESHEET_CHECK } from '../modules/local/samplesheet_check' addParams( options: params.options )
+
+workflow INPUT_CHECK {
+    take:
+    samplesheet // file: /path/to/samplesheet.tsv
+    
+    main:
+    // TODO: avoiding checking samplesheet for now, add samplesheet check later.
```
Comment (FriederikeHanssen) [inline subworkflows/input_check.nf:14]: Looks like this is implemented below already...

File: subworkflows/merge_tables_wf.nf [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+params.options = [:]
+
+include { MERGE_TABLES } from '../modules/local/merge_tables' addParams( options: params.options )
+
+workflow MERGE_TABLES_WF {
+    take:
+    tables // file: /path/to/samplesheet.tsv
+    
+    main:
+    // TODO: avoiding checking samplesheet for now, add samplesheet check later.
```
Comment (FriederikeHanssen) [inline subworkflows/merge_tables_wf.nf:14]: copy & paste from the other subworkflow?

File: subworkflows/merge_tables_wf.nf [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    tables        
+        .dump()
+        .map{it -> [ it[0].source, it[0].id, it[1] ]}
+        .groupTuple()
```
Comment (FriederikeHanssen) [inline subworkflows/merge_tables_wf.nf:18]: Apparently, it is good to provide the ```size``` of the tuple, if known. Should also speed up computation, if that is a concern here.

File: modules/local/changeo/changeo_assigngenes.nf [hunk 16]
```diff
@@ -0,0 +1,35 @@
+include { initOptions; saveFiles; getSoftwareName } from '../functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+
+process CHANGEO_ASSIGNGENES {
+    tag "$meta.id"
+
```
Comment (FriederikeHanssen) [inline modules/local/changeo/changeo_assigngenes.nf:8]: There is no resource label provided here.

**Comment by ggabernet** [comment]
yeah I'm sorry for that 🙈  DSL2 porting plus some new features at the same time, it was hard to separate one from another!

File: modules/local/changeo/changeo_buildtrees.nf [hunk 17]
```diff
@@ -0,0 +1,31 @@
+include { initOptions; saveFiles; getSoftwareName } from '../functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+
+process CHANGEO_BUILDTREES {
+    tag "$meta.id"
+
```
Comment (FriederikeHanssen) [inline modules/local/changeo/changeo_buildtrees.nf:8]: Resource label?

File: modules/local/changeo/changeo_buildtrees.nf [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    tuple val(meta), path(tab) // sequence tsv table in AIRR format
+
+    output:
+    tuple val(meta), path("*_lineages.tsv")
+
+    script:
+    def software = getSoftwareName(task.process)
+    """
+    BuildTrees.py -d ${tab} --outname ${meta.id} --log ${meta.id}.log --collapse --nproc $task.cpus --igphyml
+    """
```
Comment (FriederikeHanssen) [inline modules/local/changeo/changeo_buildtrees.nf:30]: Software version?

File: modules/local/changeo/changeo_convertdb_fasta.nf [hunk 19]
```diff
@@ -0,0 +1,30 @@
+include { initOptions; saveFiles; getSoftwareName } from '../functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+
+process CHANGEO_CONVERTDB_FASTA {
+    tag "$meta.id"
+
```
Comment (FriederikeHanssen) [inline modules/local/changeo/changeo_convertdb_fasta.nf:8]: Resource label?

File: modules/local/changeo/changeo_creategermlines.nf [hunk 20]
```diff
@@ -0,0 +1,55 @@
+include { initOptions; saveFiles; getSoftwareName } from '../functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+
+process CHANGEO_CREATEGERMLINES {
+    tag "$meta.id"
+
```
Comment (FriederikeHanssen) [inline modules/local/changeo/changeo_creategermlines.nf:8]: Resource label?

File: modules/local/changeo/changeo_creategermlines.nf [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+        ${imgt_base}/${params.species}/vdj/imgt_${params.species}_TRAJ.fasta \\
+        ${imgt_base}/${params.species}/vdj/imgt_${params.species}_TRAJ.fasta \\
+        ${imgt_base}/${params.species}/vdj/imgt_${params.species}_TRBD.fasta \\
+        ${imgt_base}/${params.species}/vdj/imgt_${params.species}_TRBJ.fasta \\
+        ${imgt_base}/${params.species}/vdj/imgt_${params.species}_TRDD.fasta \\
+        ${imgt_base}/${params.species}/vdj/imgt_${params.species}_TRDJ.fasta \\
+        ${imgt_base}/${params.species}/vdj/imgt_${params.species}_TRGJ.fasta \\
+        --log ${meta.id}.log --outname ${meta.id} > ${meta.id}_command_log.txt
+        ParseLog.py -l ${meta.id}.log -f ID V_CALL D_CALL J_CALL
+        """
```
Comment (FriederikeHanssen) [inline modules/local/changeo/changeo_creategermlines.nf:52]: software version?

File: modules/local/changeo/changeo_defineclones.nf [hunk 22]
```diff
@@ -0,0 +1,42 @@
+include { initOptions; saveFiles; getSoftwareName } from '../functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+
+process CHANGEO_DEFINECLONES {
+    tag "$meta.id"
+
```
Comment (FriederikeHanssen) [inline modules/local/changeo/changeo_defineclones.nf:8]: Resource label?

File: modules/local/changeo/changeo_makedb.nf [hunk 23]
```diff
@@ -0,0 +1,43 @@
+include { initOptions; saveFiles; getSoftwareName } from '../functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+
+process CHANGEO_MAKEDB {
+    tag "$meta.id"
+
```
Comment (FriederikeHanssen) [inline modules/local/changeo/changeo_makedb.nf:8]: Resource label?

File: modules/local/changeo/changeo_parsedb_select.nf [hunk 24]
```diff
@@ -0,0 +1,37 @@
+include { initOptions; saveFiles; getSoftwareName } from '../functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+
+process CHANGEO_PARSEDB_SELECT {
+    tag "$meta.id"
+
```
Comment (FriederikeHanssen) [inline modules/local/changeo/changeo_parsedb_select.nf:8]: Resource label?

File: modules/nf-core/software/multiqc/main.nf [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+
+process MULTIQC {
+    label 'process_medium'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename: filename, options: params.options, publish_dir: getSoftwareName(task.process), publish_id: '') }
+
+    conda (params.enable_conda ? "bioconda::multiqc=1.9" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/multiqc:1.9--pyh9f0ad1d_0"
```
Comment (apeltzer) [inline modules/nf-core/software/multiqc/main.nf:15]: Maybe use the latest version here - 1.11 is latest.

**Review by apeltzer — APPROVED** [review]
Mostly good, some points that I wasn't sure whether there is a reason for them:

- bibtex entries are not used anywhere, so could be deleted I guess?
- Bump Nextflow to 21.04.1 maybe, as this fixes some bugs in 21.04 and is also stable release 👍🏻 
- Bump MultiQC 1.9 --> 1.11, maybe fixes some bugs in 1.9 / 1.10 and / 1.10.1 too - so good if you anyways prepare for a release to include it 
- Maybe check upgrading some other packages to the respective latest versions, though I'd stay away from the R packages as they tend to break in some cases I fear

Otherwise looks really good - the above are minor bits I think 😄

**Review by FriederikeHanssen — COMMENTED** [review]
Looks awesome 🎉 , without having any knowledge of how bcell analysis works 😄 . For DSL2: Some modules are missing resource labels and/or obvious software versions (maybe they are present in the log files that are produced) . I tried to tag them, but might have missed some here and there now as well. If that is intended, then just ignore my comments there. Same for some TODO statements here and there.

The CI tests is using an older nextflow version than stated in the nextflow.config. 

The comment about ```docker.userEmulation```: Maybe someone else form the modules team knows more here. I just wanted to bring it up, because i am not sure how widely used it is now.

File: nextflow.config (hunk 12 above)
Comment (apeltzer) [reply nextflow.config:130]: Ha, never heard about that one  - we should have that in the template I guess 😓

Comment (FriederikeHanssen) [reply nextflow.config:130]: Found it: https://github.com/nf-core/tools/pull/1095 I guess it is

File: .github/workflows/ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (FriederikeHanssen) [reply .github/workflows/ci.yml:52]: To clarify , above in the file 😆 🤦‍♀️

File: assets/references.bibtex (hunk 8 above)
Comment (ggabernet) [reply assets/references.bibtex:1]: yes right, I'll update the bibtex file to include the references for this pipeline in the report, I forgot about this completely!

File: .github/workflows/ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ggabernet) [reply .github/workflows/ci.yml]: thanks for spotting this!

File: modules/local/fetch_databases.nf [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+    } else {
+        container "quay.io/biocontainers/mulled-v2-7d8e418eb73acc6a80daea8e111c94cf19a4ecfd:e5bd866a6803c301bf10b82e697c5dd2e49810a1-1"                        // Docker image
+    }
+
+    output:
+    path("igblast_base"), emit: igblast
+    path("imgtdb_base"), emit: imgt
+    
+    script:
+    //TODO: get db versions
```
[the remark this answers is not in the document]
Comment (ggabernet) [reply modules/local/fetch_databases.nf:25]: I couldn't figure out a way to get a specific DB version, but I'll include at least the date of download now.

File: bcellmagic.nf [hunk 27]
```diff
@@ -0,0 +1,384 @@
...
+    // Pre-consensus pair
+    PRESTO_PAIRSEQ (
+        PRESTO_MASKPRIMERS.out.reads
+    )
+
+    // Cluster sequences by similarity
+    PRESTO_CLUSTERSETS (
+        PRESTO_PAIRSEQ.out.reads
+    )
+    // ch_software_versions = ch_software_versions.mix(PRESTO_CLUSTERSETS.out.version.first().ifEmpty(null))
```
[the remark this answers is not in the document]
Comment (ggabernet) [reply bcellmagic.nf:196]: yes, most of the modules are part of the same software: presto, changeo, shazam or alakazam. I've grepped their versions only once, as I use always the same version.

File: modules/local/presto/presto_splitseq.nf (hunk 9 above)
Comment (ggabernet) [reply modules/local/presto/presto_splitseq.nf:31]: ah sorry, this is already fixed, as the `--fasta` option already generates the fasta file

Comment (ggabernet) [reply modules/local/presto/presto_splitseq.nf:31]: But removed the TODO now :)

File: modules/local/changeo/changeo_creategermlines.nf (hunk 20 above)
Comment (ggabernet) [reply modules/local/changeo/changeo_creategermlines.nf:8]: thanks, added them now!

File: nextflow.config (hunk 12 above)
Comment (ggabernet) [reply nextflow.config:130]: Thanks, added now!

File: modules/local/changeo/changeo_buildtrees.nf (hunk 18 above)
Comment (ggabernet) [reply modules/local/changeo/changeo_buildtrees.nf:30]: it is the changeo software, for which the version is taken under `changeo_assigngenes.nf`

File: subworkflows/merge_tables_wf.nf (hunk 14 above)
Comment (ggabernet) [reply subworkflows/merge_tables_wf.nf:14]: yup 😄

File: subworkflows/merge_tables_wf.nf (hunk 15 above)
Comment (ggabernet) [reply subworkflows/merge_tables_wf.nf:18]: It's not known a priori, it depends on the number of samples

File: assets/nf-core_style.css (hunk 7 above)
Comment (ggabernet) [reply assets/nf-core_style.css:1]: This is the css style for the Rmarkdown report 😄

**Comment by ggabernet** [comment]
Thanks a lot for your reviews @apeltzer and @FriederikeHanssen I've addressed all your points :)

**Pull request merged**